### PR TITLE
fix: Certora [M-01] ERC4626 withdrawal (SC-1212)

### PR DESCRIPTION
**Pull request by supercontracts**
## Summary by CodeRabbit

* **Tests**
  * Added and expanded ERC4626 tests covering withdrawal success/failure, zero-max-amount, and rounding edge cases; updated assertions for balances, totals, and rate-limit effects.

* **Refactor**
  * Consolidated redemption/withdrawal flow to centralize asset↔share conversion and unify rate-limit updates, improving consistency around rounding and rate-limit accounting.

### CODE REVIEW

**Review by hivtushok — DISMISSED** [review]
Looks good to me

File: src/ForeignController.sol [hunk 1]
```diff
@@ -361,19 +361,8 @@ contract ForeignController is AccessControlEnumerable {
...
-            (uint256)
-        );
-
-        rateLimits.triggerRateLimitIncrease(
-            RateLimitHelpers.makeAddressKey(LIMIT_4626_DEPOSIT, token),
-            amount
+        shares = IERC4626(token).convertToShares(
+            _redeem(token, IERC4626(token).convertToShares(amount))
         );
     }
```
Comment (lucas-manuel) [inline src/ForeignController.sol]:
```solidity
    function withdrawERC4626(address token, uint256 amount)
        external
        onlyRole(RELAYER)
        rateLimitedAddress(LIMIT_4626_WITHDRAW, token, amount)
        returns (uint256 shares)
    {
        shares = IERC4626(token).convertToShares(amount)
        _redeem(token, shares);
    }
```

File: src/ForeignController.sol [hunk 2]
```diff
@@ -383,6 +372,15 @@ contract ForeignController is AccessControlEnumerable {
         onlyRole(RELAYER)
         returns (uint256 assets)
     {
+        assets = _redeem(token, shares);
+
+        rateLimits.triggerRateLimitDecrease(
```
Comment (lucas-manuel) [inline src/ForeignController.sol]: This needs to go into _redeem

File: src/MainnetController.sol [hunk 3]
```diff
@@ -456,25 +456,20 @@ contract MainnetController is AccessControlEnumerable {
...
 
-        // Withdraw asset from a token, decode resulting shares.
-        // Assumes proxy has adequate token shares.
-        shares = abi.decode(
-            proxy.doCall(
-                token,
-                abi.encodeCall(IERC4626(token).withdraw, (amount, address(proxy), address(proxy)))
-            ),
-            (uint256)
+        shares = IERC4626(token).convertToShares(
```
Comment (lucas-manuel) [inline src/MainnetController.sol]: Same changes as Foreign

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/MainnetController.sol (hunk 3 above)
Comment (supercontracts) [reply src/MainnetController.sol]: done

File: src/ForeignController.sol (hunk 2 above)
Comment (supercontracts) [reply src/ForeignController.sol]: done

File: src/ForeignController.sol (hunk 1 above)
Comment (supercontracts) [reply src/ForeignController.sol]: done
